### ISSUE TITLE
Whitelist where `keys.evmPrivateEncrypted` can be used

### ISSUE THREAD

**Issue by 0x4007** [ubiquibot/permit-generation#20]
### Stealing Funds

We should figure out a solution for this. It is trivial to: 

1. copy the `keys.evmPrivateEncrypted` as a collaborator in an organization
2. install the bot on your own repository
3. paste in the stolen key
4. generate rewards

We should consider a whitelist system, where an encrypted key can be registered to work with an organization/repo from the same wallet that topped up the balance. This seems to be the most secure and robust way.

In this case, I imagine that we could make a tool where I would authenticate with `ubq.eth` and then I can tell the permit generation module that it has authority to generate permits in specific repositories or organizations with my `keys.evmPrivateEncrypted` found in the organization config. 

It seems a bit involved to build this out, but it is definitely something worth considering in larger organizations. I am curious to hear of proposals for how to solve this business problem more efficiently.

**Comment by 0x4007** [comment]
@rndquu rfc

**Comment by gentlementlegen** [comment]
The generation is still located within this repo, and used in several places like `conversation-rewards` for the bot V2 or `comment-incentives` for the current bot. Also this is used in `pay.ubq.fi` to decode and display permit values.

Stealing the key would not be enough, because it is encrypted, You also need the `X25519_PRIVATE_KEY` to decode it. this key is only available through our organization, so to steal it I think you would need access to the secrets. And secrets are not readable so I don't think this can be stolen.

The only way I see right now to generate permits is to trigger the `conversation-rewards` directly by triggering the action through code. Not sure if that case in handled, but I believe it should reject the run.

**Comment by 0x4007** [comment]
>  this key is only available through our organization

No its in the kernel runtime secrets, which is a single environment for everywhere the bot is added to via the GitHub App.

**Comment by rndquu** [comment]
Such kind of issues stems from the fact that private key is stored in a public repo (although encrypted). Whitelisting, restricting access, etc... will eventually lead to stolen funds due to some bug or misconfiguration. 

The most robust solution is to store private key on the partner's side but we somehow need to sign permits with partner's private key. All of the permits are saved in a DB right now. We could implement a solution when permits are generated without a signature and later partners simply adds their signatures in an automated way.

Example:
1. Permit without a signature is generated and saved in a DB
2. Partner clones some config repository, sets private key in github secrets. On CRON schedule there will be some script that will sign permits (i.e. adds signature to permits in a DB) for a particular partner using supabase access key with write access to the `permits` table.

**Comment by 0x4007** [comment]
> permits are generated without a signature and later partners simply adds their signatures in an automated way

I had no idea this is possible. 

> Partner clones some config repository, sets private key in github secrets. On CRON schedule there will be some script that will sign permits (i.e. adds signature to permits in a DB) for a particular partner using supabase access key with write access to the permits table.

This is an interesting idea. Events based would be ideal if possible. 

I am starting to think that we can find more utility for the `.ubiquibot-config` repository. We can make it have some GitHub Action that can be pinged by the kernel (just like a plugin.) Within its secrets, the partner can add an evm private key. 

The kernel can transmit some metadata that represents the source of the event context (which repository/organization) and the GitHub Action can check for this before generating a valid permit?

**Comment by rndquu** [comment]
> The kernel can transmit some metadata that represents the source of the event context (which repository/organization) and the GitHub Action can check for this before generating a valid permit?

At first glance it solves a security issue on forks. This way the `.ubiquibot-config` repository basically works in the "I can sign any incoming data" way. Even if we implement the "permit can be signed only within an org/repo where the bot is installed" feature then malicious collaborators will still be able to invoke permit signing and generate as many permits as they want. Perhaps there exists a secure solution with this approach.

The simplest solution is:
1. Partner sets private key in the `.ubiquibot-config` **repository** secrets (not organization since github secrets can be [exfiltrated](https://stackoverflow.com/questions/63003669/how-can-i-see-my-git-secrets-unencrypted) by collaborators, so ideally only partner should have read/write access to the `.ubiquibot-config` repository)
2. CRON job (with supabase write access) adds signatures for related permits

To sum up there are (at least) 2 approaches:
1. Event based (mentioned [here](https://github.com/ubiquibot/permit-generation/issues/20#issuecomment-2175273537)) when the kernel invokes the `.ubiquibot-config` repository as a plugin and asks to sign some data (requires more research if it can be implemented in a secure way)
2. Push based when permits are signed via a CRON job (mentioned [here](https://github.com/ubiquibot/permit-generation/issues/20#issuecomment-2175226503))

**Comment by rndquu** [comment]
Actually if we combine both of the approaches (event based + push based) then we get:
1. Permit is generated and saved in a DB
2. The kernel calls `.ubiquibot-config` as a plugin and says that there is a new permit in a DB to sign
3. Some script from `.ubiquibot-config` finds that permit and signs it directly in a DB

**Comment by 0x4007** [comment]
I'm really trying to not rely on databases. There's several reasons, including for reduced infrastructure complexity/maintenance; as well as that it will appeal a lot more to DAOs and other crypto natives (our primary target demographic.) It also makes development simpler to dive in and start with. 

If it is events based, we don't need to store the unsigned permit anywhere. We should transmit it directly from the kernel, to the `.ubiquibot-config` plugin.

**Comment by Keyrxng** [comment]
[more context](https://github.com/ubiquibot/permit-generation/pull/59#discussion_r1748047390)

Have we considered building EOA PK encryption around the partner's official install details or at least including it as entropy in the process?

- Only we can see installation information for an org and our bot
- Specific to each partner
- Only we can decrypt it
- Allows it to be passed around without fear
- Cannot be used by any bot installation other than the partner's making it impossible for someone to just "install" the GH app and use the stolen encrypted key
- Does mean that decryption could only ever happen from within an official UbiquityOS plugin as we do not share `APP_ID` or `APP_PRIVATE_KEY`

**Comment by 0x4007** [comment]
Replaced by https://github.com/ubiquity/ubiquibot-kernel/issues/104

**Comment by 0x4007** [comment]
If ubiquibot-config is a private repository then repository id is sufficient for a salt as per the new spec.

**Comment by Keyrxng** [comment]
> If ubiquibot-config is a private repository then repository id is sufficient for a salt as per the new spec.

Any org member can obtain that salt easily and if we want to be able to allow partners to openly store their `x25519_private_key` then using it isn't really viable. There's an open discussion on having a partner use repo/org secrets to define their `x25519_private_key` and turn the config repo into an action which handles singing permits etc which I think could be avoided with using their install info in the initial encryption (so long as only we can obtain their install info, which I believe is the case)

This also decouples the would-be config plugin from being a requirement of any payment handling plugin.

https://github.com/ubiquity/ubiquibot-kernel/issues/104 ensures the current encryption and implementation cannot be re-used but it doesn't solve the problem we have which seems to be internal org members decoding `evmPrivateEncrypted` otherwise we should already be able to pass both of these through plugin settings vs repo secrets and partner's would be able to generate their own permits.

**Comment by 0x4007** [comment]
Permit generation needs to be centralized. Taking a fee is our business model.

**Comment by Keyrxng** [comment]
> Permit generation needs to be centralized. Taking a fee is our business model.

Sorry when I said partner's generating their own permits, I meant through `permit-generation` not them generating their own on the fly. If my suggestion is valid and only we can see their install info/ID then this centralizes permits to our official bot meaning only we can decrypt their `evmPrivateKey`, not even they could once encrypted.

**Comment by whilefoo** [comment]
> [ubiquity/ubiquibot-kernel#104](https://github.com/ubiquity/ubiquibot-kernel/issues/104) ensures the current encryption and implementation cannot be re-used but it doesn't solve the problem we have which seems to be a partner's internal org members decoding `evmPrivateEncrypted`

How can partner's org members decrypt `evmPrivateEncrypted` if they don't have access to our `X25519_PRIVATE_KEY`?

**Comment by Keyrxng** [comment]
> How can partner's org members decrypt evmPrivateEncrypted if they don't have access to our X25519_PRIVATE_KEY?

They can't but an org member could if a partner stored both of those openly in the config file.

But why is our `X25519_PRIVATE_KEY` being used when it only has access to Ubiquity's funding wallet and not the partner's. I thought all partner's sign their own permits from their own funding wallet using their own `X25519_PRIVATE_KEY`? Doesn't this mean Ubiquity's funding wallet will be the only funding wallet so partners don't actually need `evmPrivateEncrypted`, unless our key is able to decrypt all of the partners `evmPrivateEncrypted`? I didn't think the plan was to centralize permits as much as that, sorry if I am running on old info.

https://github.com/ubiquibot/permit-generation/pull/59#discussion_r1748000529

> The closest thing we've come up so far with the "partners generate permits themselves" feature is https://github.com/ubiquibot/permit-generation/issues/20#issuecomment-2175340965 issue. In short:

> Partner sets his wallet private key in github repository secrets in the ubiquibot-config repository
Kernel calls ubiquibot-config as a github action plugin and gets a signed permit

**Comment by whilefoo** [comment]
> They can't but an org member could if a partner stored both of those openly in the config file I think, which is the reason we haven't implemented it yet?

That's why only we have access to the `X25519_PRIVATE_KEY`

> But why is our `X25519_PRIVATE_KEY` being used when it only has access to Ubiquity's funding wallet and not the partner's.

`X25519_PRIVATE_KEY` is not a wallet private key, it's a X25519 keypair that can encrypt/decrypt text, so we use this private key to encrypt partner's wallet private key and store it in the config. When permit generation happens we decrypt that encrypted private key from their config using `X25519_PRIVATE_KEY`.
I guess the names are a bit confusing (even for me).

**Comment by 0x4007** [comment]
> I guess the names are a bit confusing (even for me).

That's why I made it a point to always prefix the key names. For example evm private key. X25519 private key (we can consider another prefix for this one)

**Comment by Keyrxng** [comment]
> X25519_PRIVATE_KEY is not a wallet private key, it's a X25519 keypair that can encrypt/decrypt text, so we use this private key to encrypt partner's wallet private key and store it in the config. When permit generation happens we decrypt that encrypted private key from their config using X25519_PRIVATE_KEY.

Ah okay so when we encrypt their private key we use our `x25519` so that only our key is able to decrypt their `evmPrivateEncrypted`, my mistake I was running on old info as I'm sure that keygen.ubq.fi used to produce new values for both each time you ran it but I don't remember there being a masterkey used at that time. Apologies folks

**Comment by rndquu** [comment]
To sum up, right now the flow for a partner's wallet private key encryption/decryption is this one:
1. Partner encrypts his wallet private key via ubiquity's `X25519_PUBLIC_KEY` and sets this value in the `evmPrivateEncrypted` config param of the https://github.com/ubiquibot/conversation-rewards plugin
2. On permit generation ubiquity uses its own `X25519_PRIVATE_KEY` to decrypt partner's wallet private key and uses it to generate a permit on behalf of a partner

If there exists a solution for partners to generate permits themselves (in a way that partners don't expose their wallet private keys to ubiquity) and at the same time allowing the ubiquity org to [take permit fees](https://github.com/ubiquibot/permit-generation/issues/20#issuecomment-2336593325) then we could implement such solution.

**Comment by Keyrxng** [comment]
> (in a way that partners don't expose their wallet private keys to ubiquity)
I'm not sure if you mean so that our org can never actually see their key or so that it never leaves their possession?

Then I guess turning the `ubiquibot-config` into an action would be best bet for that wouldn't it as it'll never leave their org. But doesn't that mean we'd need to feed our `X25519` into their action surely in order to decrypt which could potentially be logged if they modify the action? Unless their action emits their key to us to decrypt but that exposes it to Ubiquity.

> If there exists a solution for partners to generate permits themselves (in a way that partners don't expose their wallet private keys to ubiquity) and at the same time allowing the ubiquity org to https://github.com/ubiquibot/permit-generation/issues/20#issuecomment-2336593325 then we could implement such solution.

The only real differences between the current impl and my suggestion then is:

- no 'masterkey' can directly decrypt all partner keys by itself
- decryption requires: Ubiquity' `X25519` + partner' unique install (only accessible with UbiquityOS `APP_ID` + `APP_PRIVATE_KEY`)
- partners can freely pass both of their encrypted values around without fear as only Ubiquity official plugins can decrypt
- more entropy we need to ensure is kept secure vs one 'masterkey'

But they share similarities:

- only official UbiquityOS can decrypt
- permits are centralized
- fees can be taken
- All points of failure still rest on us

**Comment by rndquu** [comment]
> I'm not sure if you mean so that our org can never actually see their key or so that it never leaves their possession?

Both. If ubiquity org doesn't have access to a partner's wallet private key it means that wallet private key doesn't leave partner's possession (i.e. repository, env variable, etc...).

> decryption requires: Ubiquity' X25519 + partner' unique install (only accessible with UbiquityOS APP_ID + APP_PRIVATE_KEY)

Each time we use a value with an entropy <128 bits we open a security hole.

Example: https://github.com/apps/ubiquibot-rndquu
App id: `285638` (19 bits entropy)
Installation id in [this](https://github.com/rndquu-org) org: `37576493` (26 bits entropy)

Now check [this](https://www.reddit.com/media?url=https%3A%2F%2Fpreview.redd.it%2Ftime-it-takes-a-hacker-to-brute-force-your-password-v0-m4iruaely8wc1.jpeg%3Fauto%3Dwebp%26s%3D6f30eeae18477fb35d5eaff0aafa0d224d0c0506) table to find out how much time it requires to brute force a number (TLDR; Github installation id: <1 minute).

Malicious user will have to brute force whatever we used for encryption (app id, installation id, etc...) + Ubiquity's `X25519_PUBLIC_KEY` to get a wallet's private key. Unless it takes >100 years it is considered insecure.

Adding additional entropies, salts or whatever only adds unnecessary complexity and (in most cases) adds a new attack vector.
